Add negative index regression test for json-patch

https://github.com/kubernetes/kubernetes/pull/120327 reverted json-patch back to 4.12.0 to fix a regression in negative index json-patch behavior

A test that exercises the bug fixed in https://github.com/kubernetes/kubernetes/pull/105896 is needed to ensure we don't accidentally bump again and lose the fix

_Originally posted by @liggitt in https://github.com/kubernetes/kubernetes/issues/120326#issuecomment-1701891191_
           
cc @Jefftree @soltysh

I saw interesting behaviour in branch `release-1.28` [helpers_test.go's TestJSONPatch function](https://github.com/kubernetes/kubernetes/blob/release-1.28/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go#L249). When I added this testcase
```go
{
			obj: &corev1.Pod{
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "test"},
				},
			},
			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
			expected: &corev1.Pod{
				TypeMeta: metav1.TypeMeta{
					Kind:       "Pod",
					APIVersion: "v1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "foo"},
				},
				Spec: corev1.PodSpec{},
			},
			expectErr: true,
}
```
I expected it to fail on negative index replace operation but it didn't and gave 
```bash
root@vandit-box:/home/vandit/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/util# go test
--- FAIL: TestJSONPatch (0.00s)
    helpers_test.go:329: testcase[3], unexpected non-error
FAIL
exit status 1
FAIL	k8s.io/kubectl/pkg/cmd/util	0.014s
```

I also ran a `make all` it also didn't fail on `release-1.28`
```bash
--- FAIL: TestJSONPatch (0.00s)
    helpers_test.go:329: testcase[3], unexpected non-error
FAIL
FAIL	k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util
```

I ran tests on master after adding 
```go
{
			obj: &corev1.Pod{
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "test"},
				},
			},
			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
			expected: &corev1.Pod{
				TypeMeta: metav1.TypeMeta{
					Kind:       "Pod",
					APIVersion: "v1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "foo"},
				},
				Spec: corev1.PodSpec{},
			},
			// expectErr: true,
}
```
And every test passed. 

Was this the desired output ? Since I am assuming the unexpected error to be false alarm or It should be like this?

I added this testcase in [rest_test.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go#L224-L246)

```go
{
	name:  "valid-negative-index-patch",
	patch: `[{"op": "replace", "path": "/spec/containers/-1/image", "value": "foo"}]`,
},
```

When I run `go test`, It fails with
```bash
--- FAIL: TestJSONPatch (0.00s)
    rest_test.go:269: valid-negative-index-patch: expect no error when applying json patch, but got the server rejected our request due to an error in our request
FAIL
exit status 1
FAIL	k8s.io/apiserver/pkg/endpoints/handlers	0.025s
```

Although When I try to patch the Pods with
```bash
kubectl patch pod nginx-deployment-86dcfdf4c6-f8p2s --type='json' -p='[{"op": "replace", "path": "/spec/containers/-1/image", "value": "hello-world"}]'
```
where the path is exactly same, it works fine. I also tried 
```
 `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`
 ```
 But the same `the server rejected our request due to an error in our request`

On further digging, It seems to me the pod doesn't contain a `container` object. (Assuming)
when I printed the pod

```&Pod{ObjectMeta:{podA      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:PodSpec{RestartPolicy:,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:,HostNetwork:false,HostPID:false,HostIPC:false,Hostname:,Subdomain:,SchedulerName:,},Status:PodStatus{Phase:,Conditions:[]PodCondition{},Message:,Reason:,HostIP:,PodIP:,StartTime:<nil>,},}```

and there's no slice present in [v1/types.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apiserver/pkg/apis/example/v1/types.go) and the slices in [meta/v1/types.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go) have `patchStrategy:"merge"`

Is there something, I'm missing ?

--- a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go
@@ -244,6 +244,10 @@ func TestJSONPatch(t *testing.T) {
 			expectedError:     "the server rejected our request due to an error in our request",
 			expectedErrorType: metav1.StatusReasonInvalid,
 		},
+		{
+			name:  "valid-negative-index-patch",
+			patch: `[{"op": "test", "value": "foo", "path": "/metadata/finalizers/-1"}]`,
+		},
 	} {
 		p := &patcher{
 			patchType:  types.JSONPatchType,
@@ -253,6 +257,7 @@ func TestJSONPatch(t *testing.T) {
 		codec := codecs.LegacyCodec(examplev1.SchemeGroupVersion)
 		pod := &examplev1.Pod{}
 		pod.Name = "podA"
+		pod.ObjectMeta.Finalizers = []string{"foo"}
 		versionedJS, err := runtime.Encode(codec, pod)
 		if err != nil {
 			t.Errorf("%s: unexpected error: %v", test.name, err)

--- a/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go
@@ -289,6 +289,26 @@ func TestJSONPatch(t *testing.T) {
 			fragment:  `[ {"op": "add", "path": "/metadata/labels/foo", "value": "bar"} ]`,
 			expectErr: true,
 		},
+		{
+			obj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "foo",
+					Finalizers: []string{"foo", "bar", "test"},
+				},
+			},
+			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
+			expected: &corev1.Pod{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Pod",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "foo",
+					Finalizers: []string{"foo", "bar", "foo"},
+				},
+				Spec: corev1.PodSpec{},
+			},
+		},
 	}
 
 	codec := runtime.NewCodec(scheme.DefaultJSONEncoder(),

--- a/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go
@@ -296,7 +296,7 @@ func TestJSONPatch(t *testing.T) {
 					Finalizers: []string{"foo", "bar", "test"},
 				},
 			},
-			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
+			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "baz"} ]`,
 			expected: &corev1.Pod{
 				TypeMeta: metav1.TypeMeta{
 					Kind:       "Pod",
@@ -304,7 +304,7 @@ func TestJSONPatch(t *testing.T) {
 				},
 				ObjectMeta: metav1.ObjectMeta{
 					Name:       "foo",
-					Finalizers: []string{"foo", "bar", "foo"},
+					Finalizers: []string{"foo", "bar", "baz"},
 				},
 				Spec: corev1.PodSpec{},
 			},

